src/jaxsim/integrators/common.py

```diff
@@ -0,0 +1,503 @@
+import jax_dataclasses
+from jax_dataclasses import Static
+
+import jaxsim.typing as jtp
+from jaxsim.utils.jaxsim_dataclass import JaxsimDataclass, Mutability
+
+try:
+    from typing import override
+except ImportError:
+    from typing_extensions import override
+
+
+# =============
+# Generic types
+# =============
+
+Time = jax.typing.ArrayLike
+TimeStep = jax.typing.ArrayLike
+State = NextState = TypeVar("State")
+StateDerivative = TypeVar("StateDerivative")
+PyTreeType = TypeVar("PyTreeType", bound=jtp.PyTree)
+
+
+class SystemDynamics(Protocol[State, StateDerivative]):
+    def __call__(
+        self, x: State, t: Time, **kwargs
+    ) -> tuple[StateDerivative, dict[str, Any]]: ...
+
+
+# =======================
+# Base integrator classes
+# =======================
+
+
+@jax_dataclasses.pytree_dataclass
+class Integrator(JaxsimDataclass, abc.ABC, Generic[State, StateDerivative]):
+
+    AuxDictDynamicsKey: ClassVar[str] = "aux_dict_dynamics"
+
+    dynamics: Static[SystemDynamics[State, StateDerivative]] = dataclasses.field(
+        repr=False, hash=False, compare=False, kw_only=True
+    )
+
+    params: dict[str, Any] = dataclasses.field(
+        default_factory=dict, repr=False, hash=False, compare=False, kw_only=True
+    )
+
+    @classmethod
+    def build(
+        cls: Type[Self], *, dynamics: SystemDynamics[State, StateDerivative], **kwargs
+    ) -> Self:
+        """
+        Build the integrator object.
+
+        Args:
+            dynamics: The system dynamics.
+            **kwargs: Additional keyword arguments to build the integrator.
+
+        Returns:
+            The integrator object.
+        """
+
+        return cls(dynamics=dynamics, **kwargs)  # noqa
+
+    def step(
+        self,
+        x0: State,
+        t0: Time,
+        dt: TimeStep,
+        *,
+        params: dict[str, Any],
+        **kwargs,
+    ) -> tuple[State, dict[str, Any]]:
+        """
+        Perform a single integration step.
+
+        Args:
+            x0: The initial state of the system.
+            t0: The initial time of the system.
+            dt: The time step of the integration.
+            params: The auxiliary dictionary of the integrator.
+            **kwargs: Additional keyword arguments.
+
+        Returns:
+            The final state of the system and the updated auxiliary dictionary.
+        """
+
+        with self.editable(validate=False) as integrator:
+            integrator.params = params
+
+        with integrator.mutable_context(mutability=Mutability.MUTABLE):
+            xf = integrator(x0, t0, dt, **kwargs)
+
+        assert Integrator.AuxDictDynamicsKey in integrator.params
+
+        return xf, integrator.params
+
+    @abc.abstractmethod
+    def __call__(self, x0: State, t0: Time, dt: TimeStep, **kwargs) -> NextState:
+        pass
+
+    def init(
+        self,
+        x0: State,
+        t0: Time,
+        dt: TimeStep,
+        *,
+        key: jax.Array | None = None,
+        **kwargs,
```

Comment:
It is not used.... now. It is there since users can create their own integrators. Maybe a random key is necessary, and here `kwargs` is used differently since it must match the `kwargs` of the `__call__` method.